### PR TITLE
add an incremental test for s2i build strategies

### PR DESCRIPTION
**Pull request by bparees**

### CODE REVIEW

**Comment by bparees** [comment]
@mfojtik @gabemontero ptal.
[test][extended:all]

File: test/extended/builds/sti_secrets.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+			err = oc.KubeFramework().WaitForPodRunning(pod.Name)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("expecting the pod container has saved artifacts")
+			out, err := oc.Run("exec").Args("-p", pod.Name, "--", "curl", "http://0.0.0.0:8080").Output()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			if !strings.Contains(out, "artifacts exist") {
+				logs, _ := oc.Run("logs").Args(pod.Name).Output()
+				e2e.Failf("Pod %q contains does not contain expected artifacts: %q\n%q", pod.Name, out, logs)
```
Comment (gabemontero) [inline test/extended/builds/sti_secrets.go]: s / contains does not contain / does not contain /

File: test/extended/builds/sti_secrets.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	g.Describe("Building from a template", func() {
+		g.It(fmt.Sprintf("should create a build from %q template and run it", templateFixture), func() {
+			oc.SetOutputDir(exutil.TestContext.OutputDir)
+
+			g.By(fmt.Sprintf("calling oc new-app -f %q", templateFixture))
+			err := oc.Run("new-app").Args("-f", templateFixture).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("starting a test build")
+			buildName, err := oc.Run("start-build").Args("initial-build").Output()
```
Comment (gabemontero) [inline test/extended/builds/sti_secrets.go]: I'm not a big fan of overloading / reusing the returned error variable, but I'm in the minority here.  I'm not calling for a change here, just interested on any additional commentary on the topic.

File: test/extended/builds/sti_secrets.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+				func(b *buildapi.Build) bool {
+					return b.Name == buildName && b.Status.Phase == buildapi.BuildPhaseComplete
+				},
+				// The build failed
+				func(b *buildapi.Build) bool {
+					if b.Name != buildName {
+						return false
+					}
+					return b.Status.Phase == buildapi.BuildPhaseFailed || b.Status.Phase == buildapi.BuildPhaseError
+				},
```
Comment (gabemontero) [inline test/extended/builds/sti_secrets.go]: Are we copying/pasting the build passed / build failed funcs each time ... I know I tried to tweak/change the version in the force pull test, but it some sort of issue, and in the rush, don't remember if I landed on the same impl that @mfojtik had for his original test that leveraged this.  If they are in fact all 3 the same now, perhaps we have enough of a trend, where we still provide this current form for flexibility, but provide a second form that defines those 2 func's internally somehow and does not require the caller to `exutil.WaitForABuild` to provide it?

File: test/extended/builds/sti_secrets.go (hunk 2 above)
Comment (bparees) [inline test/extended/builds/sti_secrets.go]: i copy+pasted most of this test from sti_env.go so i can't claim i put any thought into it.

at the end of the day i think it's mostly a style question..  there are a few places where you might save yourself by using a distinct variable but it seems relatively low risk and saves some typing.  (for that matter declaring a new variable for every error could also result in errors, if you then copy+paste a return statement and accidentally return the wrong error variable, for example)

File: test/extended/builds/sti_secrets.go (hunk 3 above)
Comment (bparees) [inline test/extended/builds/sti_secrets.go]: yup...i copy+pasted w/o studying it too carefully, but this definitely seems like a spot for a utility func "waitforbuildsuccess" or some such thing.  will look at it tomorrow.

Comment (mfojtik) [inline test/extended/builds/sti_secrets.go]: it is tricky when you want to test 'failure' scenarios, where the functions will be different. I'm not against having some 'prepared' functions that we just call here.

**Comment by bparees** [comment]
reworked to include pre-defined success/fail functions.  note that I had to do some awkward things with the build name but i don't think the end result is terrible.. the WaitForABuild method itself is now responsible for gatekeeping the buildname, and then the ok/fail functions are responsible for just looking at the build object itself.  

@mfojtik @gabemontero ptal.
[test][extended:all]

**Comment by gabemontero** [comment]
@bparees  - WaitForABuild refactor looks good to me.  There is another candidate to consume the new form of WaitForABuild in test/extended/util/image_helpers.go - I can make that change in a separate PR if you like.

**Comment by bparees** [comment]
@gabemontero thanks, missed that one.  i've gone ahead and updated it in this PR.

**Comment by bparees** [comment]
[test][extended:all]

**Comment by bparees** [comment]
@mfojtik you ok with the WaitForABuild refactor?

**Comment by bparees** [comment]
@mfojtik silence means consent.  [merge]

**Comment by mfojtik** [comment]
refactor looks good :) sorry, somehow my gmail hide this PR from me

**Pull request merged**